### PR TITLE
Implementation of new signal selection model 

### PR DESCRIPTION
**Pull request by MRD2F**

### CODE REVIEW

File: Studies/include/AnaPyInterface.h [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                    const ROOT::VecOps::RVec<LorentzVectorE>& jets_p4, const size_t& n)
+{   if(n < df_index.size())
+        return jets_p4.at(df_index.at(n)).E() / jets_p4.at(df_index.at(n)).Pt();
+    else
+        return 0;
+}
+
+
+int jet_genbJet (const ROOT::VecOps::RVec<int> jet_genJetIndex,
+                 const ROOT::VecOps::RVec<size_t>& ordered_jet_indexes, const size_t& n,
```
Comment (kandrosov) [inline Studies/include/AnaPyInterface.h]: `const size_t&` -> `size_t`

File: Studies/include/AnaPyInterface.h [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    std::sort(ordered_index.begin(), ordered_index.end(), [&](size_t a, size_t b){
+        return jets_deepFlavour.at(a) > jets_deepFlavour.at(b);
+    });
+
+    return ordered_index;
+}
+
+float httDeltaPhi_jet (const LorentzVectorM& htt_p4,const ROOT::VecOps::RVec<size_t>& df_index,
+                       const ROOT::VecOps::RVec<LorentzVectorE>& jets_p4, const size_t& n)
```
Comment (kandrosov) [inline Studies/include/AnaPyInterface.h]: `const size_t&` -> `size_t`

File: Studies/include/AnaPyInterface.h [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+{
+    if(n < df_index.size())
+        return ROOT::Math::VectorUtil::DeltaPhi(htt_p4, jets_p4.at(df_index.at(n)));
+    else
+        return 0;
+}
+
+float httDeltaEta_jet (const LorentzVectorM& htt_p4, const ROOT::VecOps::RVec<size_t>& df_index,
+                       const ROOT::VecOps::RVec<LorentzVectorE>& jets_p4, const size_t& n)
```
Comment (kandrosov) [inline Studies/include/AnaPyInterface.h]: `const size_t&` -> `size_t`

File: Studies/include/AnaPyInterface.h [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        return (htt_p4.eta() - jets_p4.at(df_index.at(n)).eta());
+    else
+        return 0;
+}
+
+
+float httMetDeltaPhi (const LorentzVectorM& htt_p4,
+                      const LorentzVectorM& met_p4)
+{
+    float dphi = ROOT::Math::VectorUtil::DeltaPhi(htt_p4, met_p4);
```
Comment (kandrosov) [inline Studies/include/AnaPyInterface.h]: this you can call directly in `df.Declare`, don't need to have a function

File: Studies/include/AnaPyInterface.h [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    return dphi;
+}
+
+float met_pt(LorentzVectorM& met_p4)
+{
+    return met_p4.pt();
```
Comment (kandrosov) [inline Studies/include/AnaPyInterface.h]: this you can call directly in `df.Declare`, don't need to have a function

File: Studies/notebooks/SignalSelectionNNTraining.ipynb [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import re \n",
+    "import os\n",
+    "import tensorflow as tf\n",
+    "import ROOT\n",
```
Comment (kandrosov) [inline Studies/notebooks/SignalSelectionNNTraining.ipynb]: ROOT is not needed

File: Studies/notebooks/SignalSelectionNNTraining.ipynb [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+    "from SignalSelectionModel import HHModel"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def ListToVector(l, elem_type):\n",
```
Comment (kandrosov) [inline Studies/notebooks/SignalSelectionNNTraining.ipynb]: Define function `FindFiles(path, pattern)` in InputsProducer and move code from this cell and the cell below there.

File: Studies/notebooks/SignalSelectionNNTraining.ipynb [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+      "2019-10-17 15:04:01.370787: W tensorflow/core/grappler/optimizers/implementation_selector.cc:310] Skipping optimization due to error while loading function libraries: Invalid argument: Functions '__inference_cudnn_lstm_with_fallback_23835_specialized_for_hh_model_lstm_1_StatefulPartitionedCall_at___inference_distributed_function_25587' and '__inference_cudnn_lstm_with_fallback_23835' both implement 'lstm_e6afa1f0-51ad-42d2-a064-c45b83d884fe' but their signatures do not match.\n"
+     ]
+    }
+   ],
+   "source": [
+    "csv_logger = CSVLogger('history.csv', append=True, separator=';')\n",
+    "history_callback = model.fit(X, Y[:,:,:], validation_split=0.2, epochs=10, batch_size=1000, callbacks=[csv_logger])"
```
Comment (kandrosov) [inline Studies/notebooks/SignalSelectionNNTraining.ipynb]: `history_callback` -> `history`

File: Studies/notebooks/SignalSelectionNNTraining.ipynb [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model.save_weights('./checkpoints/my_checkpoint')"
```
Comment (kandrosov) [inline Studies/notebooks/SignalSelectionNNTraining.ipynb]: save the entire model, not only weights.

File: Studies/python/InputsProducer.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        return all_indexes
+    elif type(x) == str :
+        idx = all_vars.index(x)
+        return idx
+
+def CreateInputs(raw_data):
+    evt_columns, jet_column, all_vars, jet_columns = CreateColums()
+    n_vars_evt = len(evt_columns)
+    n_vars_jet = len(jet_column)
+    max_jet = 10
```
Comment (kandrosov) [inline Studies/python/InputsProducer.py]: `max_jet` already defined above

File: Studies/python/InputsProducer.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        np.random.shuffle(data)
+
+    return data
+
+def CreateXY(data):
+    training_evt_vars     = [ 'sample_year', 'channelId', 'htt_pt', 'htt_eta', 'htt_phi', 'htt_met_dphi', 'rel_met_pt_htt_pt']
+    idx_training_evt_vars =  GetIndex(training_evt_vars)
+
+    training_jet_vars     = [ 'jet_{}_valid', 'jet_{}_pt', 'jet_{}_eta', 'rel_jet_{}_M_pt',
+                              'rel_jet_{}_E_pt','jet_{}_htt_deta', 'jet_{}_deepFlavour', 'jet_{}_htt_dphi' ]#, 'jet_{}_genJetIndex' ]
```
Comment (kandrosov) [inline Studies/python/InputsProducer.py]: remove commented part

File: Studies/python/InputsProducer.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
+import numpy as np
```
Comment (kandrosov) [inline Studies/python/InputsProducer.py]: add some comment in the first line

File: Studies/python/MinMaxProducer.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    n_valid = np.sum(X[:,:,val])
+    min_max = {}
+    for var,pos in var_pos.items():
+        min = np.amin(X[:, :, pos])
+        max = np.amax(X[:, :, pos])
+        min_max[var] = { 'min': truncate(float(min)), 'max': truncate(float(max)) }
+
+    f = open(output_file, 'w')
+    f.write(json.dumps(min_max, indent=4))
+    f.close()
```
Comment (kandrosov) [inline Studies/python/MinMaxProducer.py]:
```python
with open(output_file, 'w') as f:
    f.write(json.dumps(min_max, indent=4))
```

- make it runnable from the command line

File: Studies/python/SignalSelectionModel.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
+import tensorflow as tf
```
Comment (kandrosov) [inline Studies/python/SignalSelectionModel.py]: add some comment in the first line

File: Studies/python/SignalSelectionModel.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+class ScaleLayer(Layer):
+    def __init__(self, file_name, var_pos, interval_to_scale):
+        with open(file_name) as json_file:
+            data_json = json.load(json_file)
+        self.a = interval_to_scale[0]
+        self.b = interval_to_scale[1]
+        n_vars = len(var_pos)
+        self.vars_min = [0] * n_vars
+        self.vars_max = [0] * n_vars
```
Comment (kandrosov) [inline Studies/python/SignalSelectionModel.py]: better `self.vars_max = [1] * n_vars`

File: Studies/python/SignalSelectionModel.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+        self.vars_max = tf.constant(self.vars_max, dtype=tf.float32)
+        self.vars_apply = tf.constant(self.vars_apply, dtype=tf.bool)
+
+        super(ScaleLayer, self).__init__()
+
+    def build(self, input_shape):
+        super(ScaleLayer, self).build(input_shape)
+
+    def call(self, X):
+        Y = tf.clip_by_value((((self.b - self.a)*( X - self.vars_min ) / (self.vars_max - self.vars_min)) + self.a), self.a, self.b)
```
Comment (kandrosov) [inline Studies/python/SignalSelectionModel.py]: `(self.b - self.a) / (self.vars_max - self.vars_min)` can be computed once in the constructor, don't need to calc it for each call

File: Studies/python/SignalSelectionModel.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+        self.concat_1 = Concatenate()
+        self.lstm_2 = LSTM(16, return_sequences=True)
+        self.dropout_2 = Dropout(0.2)
+        self.dense_1 = TimeDistributed(Dense(10, activation="sigmoid"))
+        self.dropout_3 = Dropout(0.2)
+        self.dense_2 = TimeDistributed(Dense(1, activation="sigmoid"))
+
+    def call(self, inputs):
+        trans_x = self.normalize(inputs)
+        trans_x = self.scale(inputs)
```
Comment (kandrosov) [inline Studies/python/SignalSelectionModel.py]: `trans_x = self.scale(trans_x)`

File: Studies/python/StatProducer.py [hunk 18]
```diff
@@ -0,0 +1,25 @@
+import numpy as np
```
Comment (kandrosov) [inline Studies/python/StatProducer.py]: add some comment in the first line

File: Studies/python/Optimizer.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
+import tensorflow as tf
```
Comment (kandrosov) [inline Studies/python/Optimizer.py]: Add some comment in the first line

File: Studies/python/Optimizer.py [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+import InputsProducer
+from CalculateWeigths import ConvertToVector
+
+import ROOT
+
+parser = argparse.ArgumentParser()
+parser.add_argument("-results", "--results")
+parser.add_argument("-max", "--max_params")
+parser.add_argument("-n_iter", "--n_iter")
+parser.add_argument("-init_points", "--init_points")
```
Comment (kandrosov) [inline Studies/python/Optimizer.py]: for max, n_iter and init_points arguments add `type=int`

File: Studies/python/Optimizer.py [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+        model.build(X.shape)
+
+        history = model.fit(X, Y,sample_weight=w,validation_split=0.5, epochs=n_epoch, batch_size=params['batch_size'], verbose=0)
+        #model.save_weights()
+        tf.keras.backend.clear_session()
+        return np.amax(history.history['val_sel_acc_2'])
+    return GetLoss
+
+get_loss = CreateGetLoss(file_name, '/home/dido/CMSSW_10_2_11/src/hh-bbtautau/Studies/config/mean_std_red.json',
+                         '/home/dido/CMSSW_10_2_11/src/hh-bbtautau/Studies/config/min_max_red.json', 10)
```
Comment (kandrosov) [inline Studies/python/Optimizer.py]: don't use absolute paths

File: Studies/python/CalculateWeigths.py [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+    #sum all = n_evt
+    weights_count = np.sum(weights * counts)
+    C =  X.shape[0] / weights_count
+
+    weights *= C
+
+    return weights, mass_node_unique, mass_node
+
+# @numba.njit
+def ConvertToVector(X,Z):
```
Comment (kandrosov) [inline Studies/python/CalculateWeigths.py]: - ConvertToVector should have a numba decorator
- Create another method called `CreateSampleWeights` that will be used from outside

File: Studies/python/ParamsModel.py [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+        for var, mm in data_json.items():
+            pos = var_pos[var]
+            self.vars_min[pos] = mm['min']
+            self.vars_max[pos] = mm['max']
+            self.vars_apply[pos] = True
+        self.vars_min = tf.constant(self.vars_min, dtype=tf.float32)
+        self.vars_max = tf.constant(self.vars_max, dtype=tf.float32)
+        self.vars_apply = tf.constant(self.vars_apply, dtype=tf.bool)
+
+        # self.y = (self.b - self.a) / (self.vars_max - self.vars_min)
```
Comment (kandrosov) [inline Studies/python/ParamsModel.py]: remove comment

File: Studies/python/ParamsModel.py [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+
+        super(ScaleLayer, self).__init__()
+
+    def build(self, input_shape):
+        super(ScaleLayer, self).build(input_shape)
+
+    def call(self, X):
+        Y = tf.clip_by_value((((self.b - self.a)*( X - self.vars_min ) / (self.vars_max - self.vars_min)) + self.a), self.a, self.b)
+        # Y = tf.clip_by_value( (self.y * ( X - self.vars_min))  + self.a , self.a, self.b)
```
Comment (kandrosov) [inline Studies/python/ParamsModel.py]: remove comment

File: Studies/python/ParamsModel.py [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+
+import tensorflow as tf
+import numpy as np
+import json
+
+from tensorflow.keras.layers import Layer, Dense, Dropout, LSTM, TimeDistributed, Concatenate
+from tensorflow.keras.models import Model
+
+class StdLayer(Layer):
+    def __init__(self, file_name, var_pos, n_sigmas):
```
Comment (kandrosov) [inline Studies/python/ParamsModel.py]: add `**kwargs` and pass it to super

File: Studies/python/ParamsModel.py [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+
+    def build(self, input_shape):
+        super(StdLayer, self).build(input_shape)
+
+    def call(self, X):
+        Y = tf.clip_by_value(( X - self.vars_mean ) / self.vars_std, -self.n_sigmas, self.n_sigmas)
+        return tf.where(self.vars_apply, Y, X)
+
+class ScaleLayer(Layer):
+    def __init__(self, file_name, var_pos, interval_to_scale):
```
Comment (kandrosov) [inline Studies/python/ParamsModel.py]: add `**kwargs` and pass it to super

File: Studies/python/ParamsModel.py [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+        # Y = tf.clip_by_value( (self.y * ( X - self.vars_min))  + self.a , self.a, self.b)
+        return tf.where(self.vars_apply, Y, X)
+
+
+class HHModel(Model):
+    def __init__(self, var_pos, mean_std_json, min_max_json, params):
+        super(HHModel, self).__init__()
+
+        self.normalize = StdLayer(mean_std_json, var_pos, 5)
+        self.scale = ScaleLayer(min_max_json, var_pos, [-1,1])
```
Comment (kandrosov) [inline Studies/python/ParamsModel.py]: specify names for normalize and scale

File: Studies/python/SignalSelectionModel.py [hunk 28]
```diff
@@ -1,3 +1,6 @@
+# Definition of the model of the signal selection NN
```
Comment (kandrosov) [inline Studies/python/SignalSelectionModel.py]: remove this file from PR

File: Studies/python/ParamsModel.py (hunk 27 above)
Comment (MRD2F) [reply Studies/python/ParamsModel.py]: you mean change the names to a more accure one?

Comment (kandrosov) [reply Studies/python/ParamsModel.py]: no, I mean you should provide name argument (same way as you do for all other layers)
